library/core/src/iter/adapters/zip.rs

```diff
@@ -422,28 +422,44 @@ impl<A: Debug + TrustedRandomAccess, B: Debug + TrustedRandomAccess> ZipFmt<A, B
 ///
 /// `size` may not be overridden.
 ///
-/// `<Self as Iterator>::__iterator_get_unchecked` must be safe to call
-/// provided the following conditions are met.
+/// All subtypes and all supertypes of `Self` must also implement `TrustedRandomAccess`.
+/// In particular, this means that types with non-invariant parameters usually can not have
+/// an impl for `TrustedRandomAccess` that depends on any trait bounds on such parameters, except
+/// for bounds that come from the respective struct/enum definition itself, or bounds involving
+/// traits that themselves come with a guarantee similar to this one.
+///
+/// If `Self: Iterator`, then `<Self as Iterator>::__iterator_get_unchecked(&mut self, idx)`
+/// must be safe to call provided the following conditions are met.
 ///
 /// 1. `0 <= idx` and `idx < self.size()`.
-/// 2. If `self: !Clone`, then `get_unchecked` is never called with the same
+/// 2. If `Self: !Clone`, then `self.__iterator_get_unchecked(idx)` is never called with the same
 ///    index on `self` more than once.
-/// 3. After `self.get_unchecked(idx)` has been called then `next_back` will
-///    only be called at most `self.size() - idx - 1` times.
-/// 4. After `get_unchecked` is called, then only the following methods will be
-///    called on `self`:
-///     * `std::clone::Clone::clone()`
-///     * `std::iter::Iterator::size_hint()`
-///     * `std::iter::Iterator::next_back()`
-///     * `std::iter::Iterator::__iterator_get_unchecked()`
-///     * `std::iter::TrustedRandomAccess::size()`
+/// 3. After `self.__iterator_get_unchecked(idx)` has been called, then `self.next_back()` will
+///    only be called at most `self.size() - idx - 1` times. If `Self: Clone` and `self` is cloned,
```

Comment:
Doesn't this assume that `idx` is incremented in forwards direction? I.e. `next_back` may not do the right thing if you're calling `__iterator_get_unchecked(self.size() - 1)`

Note that the trait is called Trused**Random**Access, so theoretically some future new iterator method could take some shortcut by directly indexing into an iterator in a non-linear fashion.

Reply:
Well, first of all note that I didn’t write the original text, I just clarified it w.r.t. `Clone`, but second: if `__iterator_get_unchecked(self.size() - 1)`, then the requirement will be that
> self.next_back()` will only be called at most `self.size() - idx - 1` times

so…
> idx == size-1

> size - idx - 1 == size - (size - 1) - 1 == 0

There you go, now you’re officially not allowed to call `next_back` _at all_.

Reply:
I guess for all of this to make just a little bit more sense, there should also be the requirement (currently undocumented) that calling `next_back` decreases the size of the iterator exactly by 1.